reactioncommerce/reaction
# Add addOrderFulfillmentGroup mutation

Part of #4999
Impact: **minor**  
Type: **feature**

## Changes
A new `addOrderFulfillmentGroup` mutation allows you to add a fulfillment group to an existing order. You can provide new items for it and/or move items that are in another existing group.

The operator UI does not yet implement this.

- `addOrderFulfillmentGroup` can be called internally to synchronize from an external system by setting `context.isInternalCall` to `true`
- `addOrderFulfillmentGroup` can be called through GraphQL by any user with "orders" permission for the shop that owns the order
- All tax, shipping, surcharge, and total calculations are rerun after adjusting the existing order groups and adding the new group.

## Breaking changes
None

## Testing
1. Place an order.
1. Verify that the mutation works as described.
1. Verify that you can add a new fulfillment group when authenticated as a user with "orders" role, but not when authenticated as someone else or when not authenticated.

### Specific Instructions
1. Place an order with multiple items, and note the reference (displayed) ID.
2. Go to http://localhost:3000 and log in as an admin with "orders" permission.
3. In the browser console enter `localStorage.getItem("Meteor.loginToken")`. Copy the printed token.
4. Go to http://localhost:3000/graphql-alpha in browser
5. Expand HTTP HEADERS box and add `"meteor-login-token": "JW6NV92VfuqlrhD1M1IbBovf2ELktkhGBPeMxW6y4bj"` (the token you copied) property
6. Use the `primaryShop` query to get the primary shop ID
7. Use the following query to get your order by its reference ID and the primary shop ID

```graphql
query {
  orderByReferenceId(id: "DEfA7uGPhri4rLvhZ", shopId: "cmVhY3Rpb24vc2hvcDpKOEJocTN1VHRkZ3daeDNyeg==") {
    _id
    totalItemQuantity
    fulfillmentGroups {
      totalItemQuantity
      items {
        nodes {
          _id
          title
          variantTitle
          price {
            amount
          }
          productConfiguration {
            productId
            productVariantId
          }
        }
      }
      selectedFulfillmentOption {
        fulfillmentMethod {
          _id
        }
      }
    }
  }
}
```

8. Then use this mutation to add another group with some items and move one item from the other group:

```graphql
mutation {
  addOrderFulfillmentGroup(
    input: {
      orderId: "cmVhY3Rpb24vb3JkZXI6Zk1ySHJpRnFQRGlod21yOEM=",
      fulfillmentGroup: {
        data: {
          shippingAddress: {
            fullName: "Abe Lincoln",
            address1: "2110 Main St",
            address2: "Suite 207",
            city: "Santa Monica",
            region: "CA",
            postal: "90405",
            country: "US",
            phone: "5551234567",
            isCommercial: false
          }
        }
        selectedFulfillmentMethodId: "cmVhY3Rpb24vZnVsZmlsbG1lbnRNZXRob2Q6Nk11Z0NmM1BuNXJwZk5rZTI=",
        shopId: "cmVhY3Rpb24vc2hvcDpKOEJocTN1VHRkZ3daeDNyeg==",
        type: shipping,
        items: [{
          price: 24.99,
          quantity: 1,
          productConfiguration: {
            productId: "cmVhY3Rpb24vcHJvZHVjdDpxdHA3UmNpYXVpZ3Y2TmVSVw==",
            productVariantId: "cmVhY3Rpb24vcHJvZHVjdDpLRm12V3ZrZkxqbTNRZEdtRQ=="
          }
        },
        {
          price: 24.99,
          quantity: 1,
          productConfiguration: {
            productId: "cmVhY3Rpb24vcHJvZHVjdDpxdHA3UmNpYXVpZ3Y2TmVSVw==",
            productVariantId: "cmVhY3Rpb24vcHJvZHVjdDpLRm12V3ZrZkxqbTNRZEdtRQ=="
          }
        }]
      }
    }
  ){
    order {
      _id
      fulfillmentGroups {
        _id
        items {
          edges {
            node {
              _id
            }
          }
        }
      }
    }
  }
}
```

Be sure to copy all of the IDs from your `orderByReferenceId` query. Use a `price` that is the correct price for the product+variant combination. Use the same `selectedFulfillmentOption.fulfillmentMethod._id` as the existing group.
